### PR TITLE
fix(ivy): compile NgModule type providers without @Injectable annotation

### PR DESCRIPTION
**Pull request by pkozlowski-opensource**
Ref: FW-718

### CODE REVIEW

File: packages/upgrade/test/BUILD.bazel [hunk 1]
```diff
@@ -11,7 +11,6 @@ ts_library(
         "//packages/platform-browser",
         "//packages/platform-browser-dynamic",
         "//packages/platform-browser/testing",
-        "//packages/private/testing",
```
Comment (petebacondarwin) [inline packages/upgrade/test/BUILD.bazel]: 🥇

**Review by petebacondarwin — APPROVED** [review]
Woohoo! Last fixmeivy on upgrade??

File: packages/upgrade/test/BUILD.bazel (hunk 1 above)
Comment (pkozlowski-opensource) [reply packages/upgrade/test/BUILD.bazel]: @petebacondarwin yep, this is the last test in `ngUpgrade` :-)

**Comment by pkozlowski-opensource** [comment]
@JoostK yeh, there are more cases where a type with dependencies and without `@Inject` annotation can be used and still "work" in the current view engine (at least in the JIT mode), ex. https://stackblitz.com/edit/angular-gsutee?file=src%2Fapp%2Fapp.module.ts 

I'm referring to:

```javascript
{
      provide: 'service',
      useClass: Ng2Service,     
}
```

Need to test if it works in the AOT mode today....

**Review by mhevery — APPROVED** [review]

**Comment by mhevery** [comment]
http://test/OCL:230809548:BASE:230810729:1548375474808:e52e7cde

File: packages/core/src/di/util.ts [hunk 2]
```diff
@@ -53,3 +53,9 @@ export function convertInjectableProviderToFactory(
     return () => new type(...injectArgs(deps !));
   }
 }
+
+export function cantResolveDepsError(type: Type<any>, dependencyTokens: any[]): Error {
+  const stringifiedTokens = dependencyTokens.map(token => token ? stringify(token) : '?');
+  return new Error(
+      `Can't resolve all parameters for ${stringify(type)}: (${stringifiedTokens.join(', ')}).`);
+}
```
Comment (picheli20) [inline packages/core/src/di/util.ts]: Just a detail, the new line on the file end :)

https://github.com/angular/angular/blob/master/.editorconfig#L10

**Comment by alxhub** [comment]
(@pkozlowski-opensource and I talked about this over Slack, this is a summary of our conclusions)

While this solution would work for JIT, there's no way for the AOT compiler to support it. DI in Ivy is a runtime concept, and the AOT compiler does not introspect the providers passed to any `@NgModule`. In fact, it's impossible to do so - if you have `providers` which reference a `ModuleWithProviders` function from another module, or a constant array from another module, we don't have the information at build time.

So the correct solution here is for the user to add `@Injectable` to services they want to use via DI. This is a breaking change from View Engine in the specific cases such as this one, but the impact is expected to be minimal as most services do use `@Injectable`.

**Review by alxhub — CHANGES_REQUESTED** [review]
(see above)

**Comment by pkozlowski-opensource** [comment]
@alxhub as discussed I've removed the traversing into providers during the JIT compilation and added `@Injectable` annotation in the test to indicate that this is required now. Please take another look and approve.

**Review by alxhub — APPROVED** [review]
👍

**Pull request closed without merging**